collector/bgp.go

```diff
@@ -324,96 +335,61 @@ type bgpAdvertisedRoutes struct {
-	//       },
-	//      "10.189.0.178":{
-	// 	      "nbrDesc":"desc"
-	//      }
-	//    },
-	//    "vrf1":{
-	//      "vrfId":-1,
-	//      "vrfName":"vrf1",
-	//      "swp1":{
-	// 	      "nbrDesc":"desc"
-	//      }
-	//    }
-	// }
-	var jsonMap map[string]json.RawMessage
-	if err := json.Unmarshal(output, &jsonMap); err != nil {
-		return nil, nil, err
+func processBGPPeerDesc(output []byte) (map[string]bgpBGPNeighborsVRF, error) {
+	vrfMap := make(map[string]bgpBGPNeighborsVRF)
+	if err := json.Unmarshal([]byte(output), &vrfMap); err != nil {
+		return nil, err
 	}
+	return vrfMap, nil
+}
 
-	for vrfName, vrfData := range jsonMap {
-		var vrfFields map[string]json.RawMessage
-
-		if err := json.Unmarshal(vrfData, &vrfFields); err != nil {
-			return nil, nil, err
-		}
+func (vrf *bgpBGPNeighborsVRF) UnmarshalJSON(data []byte) error {
+	var raw map[string]*json.RawMessage
 
-		for neighbor, neighborValues := range vrfFields {
-			switch neighbor {
-			case "vrfName", "vrfId":
-				// Do nothing as we do not need the value of these fields.
-			default:
-				// All other fields are neighbors.
-				var neighborData bgpBGPNeighbor
-				if err := json.Unmarshal(neighborValues, &neighborData); err != nil {
-					return nil, nil, err
-				}
+	if err := json.Unmarshal(data, &raw); err != nil {
+		return err
+	}
 
-				if !*bgpPeerDescsText {
+	vrf.BGPNeighbors = make(map[string]bgpBGPNeighborsNeighbor)
 
-					var peerDesc map[string]string
-					if err := json.Unmarshal([]byte(neighborData.NbrDesc), &peerDesc); err != nil {
-						// Don't return an error as unmarshalling is best effort.
-						level.Error(logger).Log("msg", "cannot unmarshall bgp description", "description", neighborData.NbrDesc, "neighbor", neighbor, "err", err)
-					}
-					if _, exists := descJSON[vrfName]; !exists {
-						descJSON[vrfName] = make(map[string]map[string]string)
-					}
-					descJSON[vrfName][neighbor] = peerDesc
-				}
-				if _, exists := descText[vrfName]; !exists {
-					descText[vrfName] = make(map[string]string)
-				}
-				descText[vrfName][neighbor] = neighborData.NbrDesc
+	for k, v := range raw {
+		switch k {
+		case "vrfId":
+			if err := json.Unmarshal(*v, &vrf.ID); err != nil {
+				return err
+			}
+		case "vrfName":
+			// This is somewhat redundant, since the VRF name is a top-level key in the source JSON.
+			if err := json.Unmarshal(*v, &vrf.Name); err != nil {
+				return err
 			}
+		default:
+			var neighbor bgpBGPNeighborsNeighbor
+			if err := json.Unmarshal(*v, &neighbor); err != nil {
+				return err
+			}
+
+			vrf.BGPNeighbors[k] = neighbor
 		}
 	}
-	return descJSON, descText, nil
+	return nil
+}
+
+type bgpBGPNeighborsVRF struct {
+	ID           int                                `json:"vrfId"`
+	Name         string                             `json:"vrfName"`
+	BGPNeighbors map[string]bgpBGPNeighborsNeighbor `json:"-"`
 }
 
-type bgpBGPNeighbor struct {
-	NbrDesc string `json:"nbrDesc"`
+type bgpBGPNeighborsNeighbor struct {
```

Comment:
Is there a reason why the type and variable names stutter?

Reply:
Keeping the hierarchy of the data -- `bgpBGPNeighborsVRF` and `bgpBGPNeighborsNeighbor` both come from the `bgpNeighbors` output.

Reply:
Hmm... that practice could get a bit out of hand with some of the deeply nested structs. The Go compiler knows the hierarchy - that's one of the joys of working with statically typed languages.

Reply:
That's fair enough. Do you think `bgpNeighbor` and `bgpNeighborVRF` is suitable? If not, I'm open to suggestions.

Reply:
I think `bgpNeighbor` is fine, since it needs to be distinguishable from e.g. OSPF neighbors. However a VRF is not really a "child" of a BGP neighbor, in terms of hierarchy. If anything, it's higher up. I would name it `bgpVRF`, since it's the unmarshalled output of the `show bgp vrf all neighbors` command.

Reply:
Done.